src/main/java/org/folio/search/integration/error/KafkaErrorHandler.java

```diff
@@ -0,0 +1,58 @@
+package org.folio.search.integration.error;
+
+import static org.apache.commons.lang3.exception.ExceptionUtils.getThrowableList;
+
+import java.util.List;
+import org.apache.kafka.clients.consumer.ConsumerRecord;
+import org.folio.search.domain.dto.ResourceEventBody;
+import org.folio.search.exception.TenantNotInitializedException;
+import org.hibernate.exception.SQLGrammarException;
+import org.springframework.kafka.listener.KafkaListenerErrorHandler;
+import org.springframework.kafka.listener.ListenerExecutionFailedException;
+import org.springframework.messaging.Message;
+import org.springframework.stereotype.Component;
+
+@Component
+public class KafkaErrorHandler implements KafkaListenerErrorHandler {
+  @Override
+  public Object handleError(Message<?> message, ListenerExecutionFailedException exception) {
+    if (isTenantNotInitializedYet(exception)) {
+      throw new TenantNotInitializedException(getTenantNames(message));
+    } else {
+      throw exception;
+    }
+  }
+
+  private boolean isTenantNotInitializedYet(ListenerExecutionFailedException exception) {
+    // In case when the schema for tenant is not created
+    // DB will throw an error saying that a table does not exists
+    // which is treated as a grammar exception.
+    // It is possible to have false positives here, when an invalid SQL is used
+    // but anyway it results in retrying the consume operation
+    return getThrowableList(exception).stream().anyMatch(SQLGrammarException.class::isInstance);
+  }
+
+  @SuppressWarnings("unchecked")
+  private String[] getTenantNames(Message<?> message) {
+    if (!isSupportedType(message)) {
+      return new String[0];
```

Comment:
Curious, in this case, Is it still a `TenantNotInitializedException` even though there are no tenant names?

Reply:
Make sense, removed this part.